File: EKF.m
function [xt_est, Pt_est] = EKF(x, P, u, z, c, N, coef) %, r_coef, q_coef)
    r_coef = coef(1);
    q_coef = coef(2);
    dt = 0.02;
    Fx = [eye(3) zeros(3,3*N)];
    xt_est = x + Fx' * g(x,u,dt);
    G = Fx' * g1(x,u,dt) * Fx;
    Gt = eye(size(G)) + G;
    Rt = eye(3) * r_coef;  % 1.4963; %
    Qt = eye(3) * q_coef;  % 0.0896; %
    Pt_est = Gt * P * Gt' + Fx' * Rt *Fx;

    for i = 1:3:size(z,2)
       j = c(i,1);
       if is_new_landmark(x, j+2)
           xt_est(j:j+2) = calc_new_lmk(xt_est,z(:,i));
       end
       z_est = h(xt_est, j);
       Fxj = [Fx;zeros(3,3+3*N)];  
       Fxj(4:end,j:j+2) = eye(3);
       Hti = h1(xt_est,j) * Fxj;
       K = nanzero((Pt_est * Hti')/(Hti*Pt_est*Hti' + Qt));
       restore_lmks = xt_est(6:3:end); % in case K modifies these
       xt_est = xt_est + K * (z(:,i) - z_est);
       I = eye(size(P));
       xt_est(3) = wrap(xt_est(3)); % make sure a is btwn -pi and pi
       xt_est(6:3:end) = restore_lmks;  % landmark ids are ints
       Pt_est = (I - K*Hti) * Pt_est;
    end
end
function pose = g(x, u, dt)
    
    v = u(1);
    w = u(2);
    a = x(3);
    if w == 0  % prevent division by zero
        pose = [v * cos(a) * dt
                v * sin(a) * dt
                0];
    else
        vw = v/w;
        pose = [...
                -vw * sin(a) + vw * sin(a + w * dt)
                 vw * cos(a) - vw * cos(a + w * dt)
                 w * dt];
    end
end
function pose = g1(x, u, dt)
%TODO Work out w=0
    v = u(1);
    w = u(2);
    a = x(3);
    if w == 0  % prevent division by zero
        vw = v;
        pose = [...
            0, 0, -v * sin(a)
            0, 0, v * cos(a)
            0, 0, 0];  % Is this line correct??
    else
        vw = v/w;

        pose = [...
            0, 0, -vw * cos(a) + vw * cos(a + w * dt)
            0, 0, -vw * sin(a) + vw * sin(a + w * dt)
            0, 0,  0
           ];
    end
end
function z = h(x_est, j)
    dx = (x_est(j) - x_est(1));
    dy = (x_est(j+1) - x_est(2));
    a = x_est(3);
    d = [dx;dy]; 
    q = d' * d;
    z = [...
        sqrt(q)
        wrap(atan2(dy, dx) - a);
        x_est(j+2)];
end
function z = h1(x_est, j)
    dx = (x_est(j) - x_est(1));
    dy = (x_est(j+1) - x_est(2));
    d = [dx;dy]; 
    q = d' * d;
    if q
        k = 1/q;
    else
        k = 0;
    end
    sq = sqrt(q);
    z = k * [...
        -sq*dx, -sq*dy, 0,  sq*dx, sq*dy, 0
         dy,    -dx,   -q, -dy,    dx,    0
         0,      0,     0,  0,     0,     q];
end

function new = is_new_landmark(x, c)
    new = x(c) == 0;
end

function xj = calc_new_lmk(x ,zj)
        r = zj(1);  % range
        b = zj(2);  % bearing
        s = zj(3);  % lmk id
        a = x(3); % robot heading
        c = zj(3);
        xr = [x(1:2); c];
        xj = xr + [r*cos(a + b); r*sin(a+b); 0];
end

function x = nanzero(x)
    x(isnan(x)) = 0;
    return
end

function ao = wrap(ao)
%atan2(cos(a), sin(a))
    while ao > pi || ao < -pi
        if ao > pi
            ao = ao - 2*pi;
        end
        if ao < -pi
            ao = ao + 2*pi;
        end
    end
end

File: jsim/src/clojure/jsim/ekf.clj
(ns jsim.ekf
  (:require [core.matrix :refer :all]))

(defn predict
  [robot control]
  (let [motion-noise (:motion-noise robot)
        pose_0 (:pose robot)
        P (:covariance robot)
        move (:motion robot)
        move' (:motion' robot)
        I3 (identity-matrix 3)
        I3' (transpose I3)
        R (* motion-noise I3)
        Ig [[1 0 0 0 0 0 0]
            [0 1 0 0 0 0 0]
            [0 0 1 0 0 0 0]]
        noise-matrix (* I3' R I3 )
        G (+ Ig (move' pose_0 control))
        G' (transpose G)]
    {:x (move pose_0 control)
     :P (+ (* G P G') R)}))

(defn- new-landmark?
  [landmark feature-map]
  (let [id (nth landmark 2)
        ids (take-nth 3 (drop 2 feature-map))]
    (contains? (set ids) 2)))

(defn- update-landmark ;; What are my assumptions here?
  [landmark observe observe' x p q]
  (let [Ih (identity-matrix 6)
        z (observe x landmark) ; Observation in global grame
        H (* (observe' x landmark) Ih) ; Linearized observation
        H' (transpose H)
        Q (* I3 q)
        P_est (* p H')
        info-matrix (+ (* H Pt_est H') Q)
        K (/ P_est info-matrix); Kalman Gain - Does this just affect the observed landmark??
        z- (- landmark z)
        ]
    (if (new-landmark? landmark)
      ("do a")
      ("do b"))
    {:K K
     :z-diff z-}))

(defn update
  [robot
   observations]
  (let [sensor-noise (:sensor-noise robot)
        pose (:pose robot)
        observe (:observation robot)
        feature-map (:feature-map robot)]
    (if new-landmark? )) )

(defn ekf
  [robot
   control
   observations
   update' #(update %1 observations)]
  (->
   (predict robot control)
   update'))


;    for i = 1:3:size(z,2)
;       j = c(i,1);
;       if is_new_landmark(x, j+2)
;           xt_est(j:j+2) = calc_new_lmk(xt_est,z(:,i));
;       end
;       z_est = h(xt_est, j);
;       Fxj =   [Fx;zeros(3,3+3*N)];
;       Fxj(4:end,j:j+2) = eye(3);
;       Hti = h1(xt_est,j) * Fxj;
;       K = nanzero((Pt_est * Hti')/(Hti*Pt_est*Hti' + Qt));
;       restore_lmks = xt_est(6:3:end); % in case K modifies these
;       xt_est = xt_est + K * (z(:,i) - z_est);
;       I = eye(size(P));
;       xt_est(3) = wrap(xt_est(3)); % make sure a is btwn -pi and pi
;       xt_est(6:3:end) = restore_lmks;  % landmark ids are ints
;       Pt_est = (I - K*Hti) * Pt_est;
;    end
;end
